java/src/main/java/com/google/protobuf/ByteBufferByteString.java

```diff
@@ -0,0 +1,417 @@
+    } else {
+      bytes = new byte[buffer.remaining()];
+      offset = 0;
+      copyToInternal(bytes, buffer.position(), offset, size());
+    }
+    return new String(bytes, offset, size(), charset);
+  }
+
+  @Override
+  public boolean isValidUtf8() {
+    byte[] bytes;
+    int startIndex;
+    if (buffer.hasArray()) {
+      bytes = buffer.array();
+      startIndex = buffer.arrayOffset() + buffer.position();
+    } else {
+      bytes = new byte[buffer.remaining()];
+      startIndex = 0;
+      copyToInternal(bytes, 0, 0, size());
+    }
+    return Utf8.isValidUtf8(bytes, startIndex, startIndex + size());
+  }
+
+  @Override
+  protected int partialIsValidUtf8(int state, int offset, int length) {
+    byte[] bytes;
+    int startIndex;
+    if (buffer.hasArray()) {
+      bytes = buffer.array();
+      startIndex = buffer.arrayOffset() + buffer.position();
+    } else {
+      bytes = new byte[buffer.remaining()];
+      startIndex = 0;
+      copyToInternal(bytes, 0, 0, size());
+    }
+    return Utf8.partialIsValidUtf8(state, bytes, startIndex, startIndex + size());
+  }
+
+  @Override
+  public boolean equals(Object other) {
+    if (other == this) {
+      return true;
+    }
+    if (!(other instanceof ByteString)) {
+      return false;
+    }
+    ByteString otherString = ((ByteString) other);
+    if (size() != otherString.size()) {
+      return false;
+    }
+    if (size() == 0) {
+      return true;
+    }
+    if (other instanceof ByteBufferByteString) {
+      return buffer.equals(((ByteBufferByteString) other).buffer);
+    }
+    if (other instanceof RopeByteString) {
+      return other.equals(this);
+    }
+    if (other instanceof LiteralByteString) {
+      LiteralByteString lbs = (LiteralByteString) other;
+      ByteBuffer otherByteBuffer = ByteBuffer.wrap(lbs.bytes, lbs.getOffsetIntoBytes(), lbs.size());
+      return buffer.equals(otherByteBuffer);
+    }
+    for(int index=0; index < size(); ++index) {
+      if (byteAt(index) != otherString.byteAt(index)) {
+        return false;
+      }
+    }
+    return false;
+  }
+
+  @Override
+  public int hashCode() {
+    int h = hash;
+
+    if (h == 0) {
+      int size = size();
+      h = partialHash(size, 0, size);
+      if (h == 0) {
+        h = 1;
+      }
+      hash = h;
+    }
+    return h;
+  }
+
+  @Override
+  protected int partialHash(int h, int offset, int length) {
+    for (int i = offset; i < offset + length; i++) {
+      h = h * 31 + buffer.get(i);
+    }
+    return h;
+  }
+
+  @Override
+  public InputStream newInput() {
+    return new InputStream() {
+      private final ByteBuffer buf = buffer.slice();
+      private int mark;
+
+      @Override
+      public synchronized void mark(int readlimit) {
+        mark = buf.position();
+      }
+
+      @Override
+      public boolean markSupported() {
+        return true;
+      }
+
+      @Override
+      public synchronized void reset() throws IOException {
```

Comment:
This was automatically generated since the base class declares them as synchronized.  Since the rest of the class isn't synchronized, removed.